src/compiler/checker.rs

```diff
@@ -0,0 +1,363 @@
+        },
+        _ => panic!("Expected IntrinsicType"),
+    }
+    type_checker.regular_true_type = Some(regular_true_type);
+    type_checker.number_or_big_int_type = Some(
+        type_checker
+            .get_union_type(vec![
+                type_checker.number_type().clone(),
+                type_checker.bigint_type().clone(),
+            ])
+            .into(),
+    );
+    type_checker.initialize_type_checker(host);
+    type_checker
+}
+
+impl TypeChecker {
+    fn number_type(&self) -> Rc<Type> {
+        self.number_type.as_ref().unwrap().clone()
+    }
+
+    fn bigint_type(&self) -> Rc<Type> {
+        self.bigint_type.as_ref().unwrap().clone()
+    }
+
+    fn true_type(&self) -> Rc<Type> {
+        self.true_type.as_ref().unwrap().clone()
+    }
+
+    fn number_or_big_int_type(&self) -> Rc<Type> {
+        self.number_or_big_int_type.as_ref().unwrap().clone()
+    }
+
+    fn diagnostics(&self) -> &RwLock<DiagnosticCollection> {
+        &self.diagnostics
+    }
+
+    fn create_error<TNode: NodeInterface>(
+        &self,
+        location: Option<&TNode>,
+        message: &DiagnosticMessage,
+    ) -> Rc<Diagnostic> {
+        if let Some(location) = location {
+            Rc::new(create_diagnostic_for_node(location, message).into())
+        } else {
+            unimplemented!()
+        }
+    }
+
+    fn error<TNode: NodeInterface>(
+        &self,
+        location: Option<&TNode>,
+        message: &DiagnosticMessage,
+    ) -> Rc<Diagnostic> {
+        let diagnostic = self.create_error(location, message);
+        self.diagnostics().write().unwrap().add(diagnostic.clone());
+        diagnostic
+    }
+
+    fn error_and_maybe_suggest_await<TNode: NodeInterface>(
+        &self,
+        location: &TNode,
+        message: &DiagnosticMessage,
+    ) -> Rc<Diagnostic> {
+        let diagnostic = self.error(Some(location), message);
+        diagnostic
+    }
+
+    fn create_type(&self, flags: TypeFlags) -> BaseType {
+        let result = (self.Type)(flags);
+        result
+    }
+
+    fn create_intrinsic_type(&self, kind: TypeFlags, intrinsic_name: &str) -> BaseIntrinsicType {
+        let type_ = self.create_type(kind);
+        let type_ = BaseIntrinsicType::new(type_);
+        type_
+    }
+
+    fn add_type_to_union(&self, type_set: &mut Vec<Rc<Type>>, type_: Rc<Type>) {
+        type_set.push(type_);
+    }
+
+    fn add_types_to_union(&self, type_set: &mut Vec<Rc<Type>>, types: &[Rc<Type>]) {
+        for type_ in types {
+            self.add_type_to_union(type_set, type_.clone());
+        }
+    }
+
+    fn get_union_type(&self, types: Vec<Rc<Type>>) -> Rc<Type> {
+        let mut type_set: Vec<Rc<Type>> = vec![];
+        self.add_types_to_union(&mut type_set, /*TypeFlags::empty(), */ &types);
+        self.get_union_type_from_sorted_list(type_set)
+    }
+
+    fn get_union_type_from_sorted_list(&self, types: Vec<Rc<Type>>) -> Rc<Type> {
```

Comment:
Maybe I should make eg `SortedList` a concrete type so that I can enforce at the type level that I'm passing an already-sorted list since it seems to be distinguishing those (though I haven't implemented any of the actual sorting logic yet or corresponding eg insertion/lookup optimizations)

Reply:
I like that idea

Reply:
Per other comment, there is in fact a `SortedArray` type but for whatever reason the Typescript version of this helper accepts a plain old `Type[]` rather than a `SortedArray` 🤷‍♀️